sessionctx/variable/session.go

```diff
@@ -1440,8 +1440,13 @@ func (s *SessionVars) IsMPPEnforced() bool {
 // TODO: Confirm whether this function will be inlined and
 // omit the overhead of string construction when calling with false condition.
 func (s *SessionVars) RaiseWarningWhenMPPEnforced(warning string) {
-	if s.IsMPPEnforced() && s.StmtCtx.InExplainStmt {
+	if !s.IsMPPEnforced() {
+		return
+	}
+	if s.StmtCtx.InExplainStmt {
 		s.StmtCtx.AppendWarning(errors.New(warning))
+	} else {
+		s.StmtCtx.AppendExtraWarning(errors.New(warning))
```

Comment:
Generating diagnostic strings may bring some overhead to the compile phase. The potential overhead is worth it for enhanced diagnosability. Maybe we can test the potential overhead after  #39893 is done(check whether compile duration increases in some typical workloads).

Reply:
You are right.
Though usually there would not be too many such strings, I'll run a benchmark then.